File: src/main/java/dao/OrderDao.java
package dao;

import com.fasterxml.jackson.core.JsonProcessingException;

import model.Order;
import model.ValidationErrors;

import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.transaction.Transactional;
import javax.validation.ConstraintViolation;
import javax.validation.Validation;
import java.util.List;


@Repository
public class OrderDao {

    @PersistenceContext
    private EntityManager em;


    public Order findById(Long idToFind) {
        return em.find(Order.class, idToFind);
    }

    public List<Order> findAll() {
        return em.createQuery("SELECT o FROM Order o", Order.class).getResultList();
    }

    @Transactional
    public Order save(Order order) {
        if (order.getId() == null) {
            em.persist(order);
        } else {
            em.merge(order);
        }
        return order;
    }

    @Transactional
    public void delete(Long id) {
        Order order = findById(id);
        if (order != null) {
            em.remove(order);
        }
    }

    public ValidationErrors validate(Order order) throws JsonProcessingException {
        var validator = Validation.buildDefaultValidatorFactory().getValidator();

        var violations = validator.validate(order);

        ValidationErrors errors = new ValidationErrors();

        for (ConstraintViolation<Order> violation : violations) {
            errors.addErrorMessage(violation.getMessage());
        }

        return errors;
    }


}

File: src/main/kotlin/model/OrderDao.kt
package model

import org.springframework.jdbc.core.JdbcTemplate
import app.OrderRowHandler
import org.springframework.jdbc.core.simple.SimpleJdbcInsert
import kotlin.Throws
import com.fasterxml.jackson.core.JsonProcessingException
import org.springframework.stereotype.Repository
import java.io.Serializable
import javax.validation.Validation

@Repository
class OrderDao(var template: JdbcTemplate) {
    fun findOrderById(idToFind: Long?): Order {
        val sql = "SELECT * FROM orders LEFT JOIN orderlines ON orders.id = orderlines.orderid WHERE orders.id = ?"
        val handler = OrderRowHandler()
        template.query(sql, handler, idToFind)
        return handler.getResult()[0]
    }

    fun findAllOrders(): List<Order> {
        val sql = "SELECT * FROM orders LEFT JOIN orderlines ON orders.id = orderlines.orderid"
        val handler = OrderRowHandler()
        template.query(sql, handler)
        return handler.getResult()
    }

    fun saveOrder(order: Order): Order {
        val newOrder = addNewOrder(order)
        if (newOrder.orderRows != null) {
            addNewOrderRows(newOrder)
        }
        return newOrder
    }

    private fun addNewOrder(order: Order): Order {
        val data = java.util.Map.of("ordernumber", order.orderNumber)
        val number = SimpleJdbcInsert(template)
            .withTableName("orders")
            .usingGeneratedKeyColumns("id")
            .executeAndReturnKey(data)
        order.id = number.toLong()
        return order
    }

    private fun addNewOrderRows(order: Order) {
        for (row in order.orderRows!!) {
            val data2: Map<String, Serializable?> = java.util.Map.of(
                "itemName", row.itemName,
                "quantity", row.quantity,
                "price", row.price,
                "orderId", order.id
            )
            val rowNumber = SimpleJdbcInsert(template)
                .withTableName("orderlines")
                .usingGeneratedKeyColumns("orderLineId")
                .executeAndReturnKey(data2)
            row.id = rowNumber.toLong()
        }
    }

    fun deleteOrder(id: Long?) {
        val sql = "DELETE FROM orders WHERE id = (?)"
        template.update(sql, id)
    }

    @Throws(JsonProcessingException::class)
    fun validate(order: Order): ValidationErrors {
        val validator = Validation.buildDefaultValidatorFactory().validator
        val violations = validator.validate(order)
        val errors = ValidationErrors()
        for (violation in violations) {
            errors.addErrorMessage(violation.message)
        }
        return errors
    }


}